### PR TITLE
Elementizes cult halos + cult eyes + fixes some issues with them

### LINKED ISSUE #44146: cult nonhuman mobs dont get the cult halo overlay
this means you can have ape cultists in stealth mode which is probably not intentional

### LINKED ISSUE #57204: Cult changelings lose their halo and glowy eyes when they transform
## Round ID:
#157131

## Reproduction:

When a cultist changeling transforms, they lose their halo and eyes. When they transform from a cultist dna to a noncultist dna back to a cultist dna, they don't seem to regain their eyes or halos.

### LINKED ISSUE #58585: Cultist Halo vanishes
## Reproduction:
I was transformed into a monkey as cultist and after my return to humanity i lacked the halo and my glowing eyes.

### PR DESCRIPTION
**Pull request by SuperNovaa41**
## About The Pull Request
Fixes https://github.com/tgstation/tgstation/issues/58585
Fixes https://github.com/tgstation/tgstation/issues/57204
Fixes https://github.com/tgstation/tgstation/issues/44146
Turns cult eyes and cult halos into separate elements.
The adding and removing of red eyes and halos are handled by said elements.

please i think this looks pretty good 

## Why It's Good For The Game
Refactors cult code a bit, makes it far easier to manage the overlays and traits given by the size of the cult.
Fixes some exploits and issues.

## Changelog

:cl:
fix: You can no longer delete your cult eyes or halo by transforming into a monkey, and then into a human.
fix: You can no longer delete your cult eyes or halo by transforming as a changeling.
fix: Simplemobs now receive glowing eyes and a cult halo.
refactor: Turns cult eyes and cult halos into elements, and improves the code a touch.
/:cl:

### CODE REVIEW

**Comment by SuperNovaa41** [comment]
last one i promise

File: code/__DEFINES/dcs/signals/signals_mob/signals_mob_carbon.dm [hunk 1]
```diff
@@ -79,3 +79,12 @@
...
 #define COMSIG_HUMAN_CHECK_SHIELDS "human_check_shields"
 	#define SHIELD_BLOCK (1<<0)
+
+// Mob transformation signals
+///Human transforming into a monkey
+#define COMSIG_HUMAN_MONKEYIZE "human_monkeyize"
+///Monkey transforming into a human
+#define COMSIG_MONKEY_HUMANIZE "monkey_humanize"
+///Changeling transforming
+#define COMSIG_CHANGELING_TRANSFORM "changeling_transform"
```
Comment (Mothblocks) [inline code/__DEFINES/dcs/signals/signals_mob/signals_mob_carbon.dm:89]: This should be in its own file (signals_changeling or similar), and be clear on what inputs its taking and when this happens

File: code/datums/elements/cult_eyes.dm [hunk 2]
```diff
@@ -0,0 +1,56 @@
+/**
+ * # Cult eyes element
+ *
+ * Applies and removes the glowing cult eyes
+ */
+/datum/element/cult_eyes
```
Comment (Mothblocks) [inline code/datums/elements/cult_eyes.dm:6]: You need `element_flags = ELEMENT_DETACH`

File: code/datums/elements/cult_eyes.dm [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * Cult eye setter proc
+ *
+ * Changes the eye color, and adds the glowing eye trait to the mob.
+ */
+/datum/element/cult_eyes/proc/set_eyes(datum/target)
+	SIGNAL_HANDLER
+
+	var/mob/living/parent_mob = target
```
Comment (Mothblocks) [inline code/datums/elements/cult_eyes.dm:31]: Just do `mob/living/target` in set_eyes definition

File: code/datums/elements/cult_eyes.dm [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+		human_parent.dna.update_ui_block(DNA_EYE_COLOR_BLOCK)
+		human_parent.update_body()
+
+/**
+ * Detach proc
+ *
+ * Removes the eye color, and trait from the mob
+ */
+/datum/element/cult_eyes/Detach(datum/target, ...)
+	. = ..()
```
Comment (Mothblocks) [inline code/datums/elements/cult_eyes.dm:45]: Call this at the end, not at the beginning

File: code/datums/elements/cult_eyes.dm [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	var/mob/living/parent_mob = target
```
Comment (Mothblocks) [inline code/datums/elements/cult_eyes.dm:46]: Same here

File: code/datums/elements/cult_eyes.dm [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	REMOVE_TRAIT(parent_mob, TRAIT_UNNATURAL_RED_GLOWY_EYES, CULT_TRAIT)
+	if (ishuman(parent_mob))
+		var/mob/living/carbon/human/human_parent = parent_mob
+		human_parent.eye_color = initial(human_parent.eye_color)
+		human_parent.dna.update_ui_block(DNA_EYE_COLOR_BLOCK)
+		human_parent.update_body()
+
+	UnregisterSignal(target, COMSIG_CHANGELING_TRANSFORM)
+	UnregisterSignal(target, COMSIG_HUMAN_MONKEYIZE)
+	UnregisterSignal(target, COMSIG_MONKEY_HUMANIZE)
```
Comment (Mothblocks) [inline code/datums/elements/cult_eyes.dm:56]: Use `UnregisterSignal(target, list(signals, go, here))` in stead

File: code/datums/elements/cult_eyes.dm [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+/datum/element/cult_eyes/Attach(datum/target, override = FALSE)
+	. = ..()
+	if (!isliving(target))
+		return ELEMENT_INCOMPATIBLE
+
+	// Register signals for mob transformation to prevent premature halo removal
+	RegisterSignal(target, COMSIG_CHANGELING_TRANSFORM, .proc/set_eyes)
+	RegisterSignal(target, COMSIG_MONKEY_HUMANIZE, .proc/set_eyes)
+	RegisterSignal(target, COMSIG_HUMAN_MONKEYIZE, .proc/set_eyes)
```
Comment (Mothblocks) [inline code/datums/elements/cult_eyes.dm:16]: Register the signals as a list, which I'm fairly certain is allowed

File: code/datums/elements/cult_eyes.dm [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+	if (override)
```
Comment (Mothblocks) [inline code/datums/elements/cult_eyes.dm:18]: This should instead be `initial_delay`, defaulting to `20 SECONDS`, but able to be set to zero.

File: code/datums/elements/cult_halo.dm [hunk 9]
```diff
@@ -0,0 +1,61 @@
+/**
+ * # Cult halo element
+ *
+ * Applies and removes the cult halo
+ */
+/datum/element/cult_halo
+
+/datum/element/cult_halo/Attach(datum/target, override = FALSE)
```
Comment (Mothblocks) [inline code/datums/elements/cult_halo.dm:8]: Same here

File: code/datums/elements/cult_halo.dm [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	. = ..()
+	if (!isliving(target))
+		return ELEMENT_INCOMPATIBLE
+
+	// Register signals for mob transformation to prevent premature halo removal
+	RegisterSignal(target, COMSIG_CHANGELING_TRANSFORM, .proc/set_halo)
+	RegisterSignal(target, COMSIG_MONKEY_HUMANIZE, .proc/set_halo)
+	RegisterSignal(target, COMSIG_HUMAN_MONKEYIZE, .proc/set_halo)
```
Comment (Mothblocks) [inline code/datums/elements/cult_halo.dm:16]: All the same here

File: code/datums/elements/cult_halo.dm [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+	if (ishuman(parent_mob))
+		var/mob/living/carbon/human/human_parent = parent_mob
+		human_parent.remove_overlay(HALO_LAYER)
+		human_parent.update_body()
+	else
+		parent_mob.cut_overlay(HALO_LAYER)
+
+	UnregisterSignal(target, COMSIG_CHANGELING_TRANSFORM)
+	UnregisterSignal(target, COMSIG_HUMAN_MONKEYIZE)
+	UnregisterSignal(target, COMSIG_MONKEY_HUMANIZE)
```
Comment (Mothblocks) [inline code/datums/elements/cult_halo.dm:61]: Same

File: code/datums/elements/cult_halo.dm [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * Halo setter proc
+ *
+ * Adds the cult halo overlays, and adds the halo trait to the mob.
+ */
+/datum/element/cult_halo/proc/set_halo(datum/target)
+	SIGNAL_HANDLER
+
+	var/mob/living/parent_mob = target
```
Comment (Mothblocks) [inline code/datums/elements/cult_halo.dm:31]: Same

File: code/datums/elements/cult_halo.dm [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+		parent_mob.add_overlay(new_halo_overlay)
+
+/**
+ * Detach proc
+ *
+ * Removes the halo overlays, and trait from the mob
+ */
+/datum/element/cult_halo/Detach(datum/target, ...)
+	. = ..()
+	var/mob/living/parent_mob = target
```
Comment (Mothblocks) [inline code/datums/elements/cult_halo.dm:50]: Same

File: code/datums/elements/cult_halo.dm [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * Adds the cult halo overlays, and adds the halo trait to the mob.
+ */
+/datum/element/cult_halo/proc/set_halo(datum/target)
+	SIGNAL_HANDLER
+
+	var/mob/living/parent_mob = target
+	ADD_TRAIT(parent_mob, TRAIT_CULT_HALO, CULT_TRAIT)
+	var/icon_state = pick("halo1", "halo2", "halo3", "halo4", "halo5", "halo6")
+	var/mutable_appearance/new_halo_overlay = mutable_appearance('icons/effects/32x64.dmi', icon_state, -HALO_LAYER)
```
Comment (Mothblocks) [inline code/datums/elements/cult_halo.dm:34]:
```suggestion
	var/mutable_appearance/new_halo_overlay = mutable_appearance('icons/effects/32x64.dmi', "halo[rand(1, 6)]", -HALO_LAYER)
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/datums/elements/cult_eyes.dm [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+ */
+/datum/element/cult_eyes/Detach(mob/living/target, ...)
+	REMOVE_TRAIT(target, TRAIT_UNNATURAL_RED_GLOWY_EYES, CULT_TRAIT)
+	if (ishuman(target))
+		var/mob/living/carbon/human/human_parent = target
+		human_parent.eye_color = initial(human_parent.eye_color)
+		human_parent.dna.update_ui_block(DNA_EYE_COLOR_BLOCK)
+		human_parent.update_body()
+	UnregisterSignal(target, list(COMSIG_CHANGELING_TRANSFORM, COMSIG_HUMAN_MONKEYIZE, COMSIG_MONKEY_HUMANIZE))
+	. = ..()
```
Comment (MrMelbert) [inline code/datums/elements/cult_eyes.dm:46]:
```suggestion
	return ..()
```
(imo) more clearer return

File: code/datums/elements/cult_halo.dm [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+/datum/element/cult_halo/Detach(mob/living/target, ...)
+	REMOVE_TRAIT(target, TRAIT_CULT_HALO, CULT_TRAIT)
+	if (ishuman(target))
+		var/mob/living/carbon/human/human_parent = target
+		human_parent.remove_overlay(HALO_LAYER)
+		human_parent.update_body()
+	else
+		target.cut_overlay(HALO_LAYER)
+	UnregisterSignal(target, list(COMSIG_CHANGELING_TRANSFORM, COMSIG_HUMAN_MONKEYIZE, COMSIG_MONKEY_HUMANIZE))
+	. = ..()
```
Comment (MrMelbert) [inline code/datums/elements/cult_halo.dm:50]:
```suggestion
	return ..()
```
ditto

**Review by MrMelbert — COMMENTED** [review]
I like this implementation much more

**Review by MrMelbert — APPROVED** [review]

**Pull request merged**